### DISCUSSION TITLE
How do I draw a spiral? 🌀

### DISCUSSION THREAD

**Discussion by Wulfsta** [libfive/libfive#549] [Q&A]
Perhaps I am thinking about this the wrong way, but I can’t come up with a way to produce a spiral. I know there is an implicit form for the Archimedean spiral, but I only want one arm, and using that form results in two. Let’s assume for simplicity that I want to sweep a sphere along the spiral path.

**Comment by omgitsraven** [comment]
It's definitely not an 'exact' SDF, and there's probably more correct *and* more efficient ways of doing it, but just as a starting point, I've gotten it done like this:

```
pi = 355.0/113.0

def spiralize(shape):
    ang = Shape.atan2(Shape.Y(),Shape.X())
    dst = Shape.sqrt(Shape.X()*Shape.X() + Shape.Y()*Shape.Y())
    dst = dst%1.0
    dst -= (ang/(pi*2))-1
    return shape.remap(
        Shape.cos(ang)*dst,
        Shape.sin(ang)*dst,
        Shape.Z()
    )

tubeThick = 0.5

spiralize(
    union(
        torus(1,tubeThick),
        torus(2,tubeThick)
    )
)
```

to give a result like this:
![image](https://github.com/libfive/libfive/assets/8540042/d29d6217-53b4-47ac-bc4e-c25b213c7249)
or this if the tubeThick is 0.25:
![image](https://github.com/libfive/libfive/assets/8540042/259aafef-9b4d-4079-bf00-ea5b85ca6e86)

The second torus is necessary to smooth over the discontinuity where the modulo jumps it up a level....  I'm sure there's a smarter way to handle that, but I usually just settle for sticking a second shape into the gap manually...

You can play with the values to eliminate that little curl towards the center if that's not what you want (maybe just by cutting it short and unioning in a sphere at the cutoff point), but hopefully this is enough for an example...

**Reply by Wulfsta** [reply]
This is very cool, and I need to spend some time wrapping my head around it. Interesting that at `tubeThick = 0.25` it divides the x,y-plane in the same way that an Archimedean spiral does.

**Reply by Wulfsta** [reply]
Your use of `atan2` got me thinking, and I’ve arrived at `f(x, y) = cos((x^2 + y^2)^(1/2) - atan2(y, x)) - 1`. I’ll have to play around with it later.

**Comment by Wulfsta** [comment]
@omgitsraven here is one for distance from the involute spiral:
```
import studio
studio.set_bounds([-50, -50, -10], [50, 50, 10])
studio.set_quality(8)
studio.set_resolution(20)

from libfive.stdlib import *

pi = 3.14159265358979

def heaviside_step_function(a):
    return Shape.min(0.0, Shape.compare(a, 0.0)) + 1.0

def if_nonnegative_else(a, f_zero, f_one):
    res = heaviside_step_function(a)
    return res * f_zero + (1.0 - res) * f_one

def planar_euclidian_distance(a, b):
    return Shape.sqrt(a * a + b * b)

def euclidian_distance(a, b, c):
    return Shape.sqrt(a * a + b * b + c * c)

def floor_shape(a):
    return a - (a % 1.0)

def round_shape(a):
    return floor_shape(a + 0.5)

def involute_nearest(x, y, circ_radius=1.0):
    theta_pre = Shape.atan2(y, x)
    theta = if_nonnegative_else(theta_pre, theta_pre, theta_pre + 2.0 * pi)
    r = planar_euclidian_distance(x, y)
    phi_pre = Shape.acos(circ_radius / r) + theta
    phi = if_nonnegative_else(phi_pre - 2.0 * pi, phi_pre - 2.0 * pi, phi_pre)
    hypotenuse_dist = planar_euclidian_distance(r, circ_radius)
    first_arm_dist = circ_radius * phi
    integer_component = Shape.max(0, round_shape((hypotenuse_dist - first_arm_dist) / (2 * pi * circ_radius)))
    nearest_arm_dist = first_arm_dist + integer_component * 2 * pi * circ_radius
    r_nearest = planar_euclidian_distance(circ_radius, nearest_arm_dist)
    theta_nearest = phi - Shape.acos(circ_radius / r_nearest)
    x_nearest = r_nearest * Shape.cos(theta_nearest)
    y_nearest = r_nearest * Shape.sin(theta_nearest)
    return (x_nearest, y_nearest)

def disk(r):
    return planar_euclidian_distance(Shape.X(), Shape.Y()) - r

def planar_involute(r, s, eps=0.1):
    x = Shape.X()
    y = Shape.Y()
    x_nearest, y_nearest = involute_nearest(x, y, r)
    x_nearest = Shape.nanfill(x_nearest, r)
    y_nearest = Shape.nanfill(y_nearest, 0)
    involute_dist_default = planar_euclidian_distance(Shape.X() - x_nearest, Shape.Y() - y_nearest)
    involute_dist = if_nonnegative_else(involute_dist_default - pi * r, Shape.abs(involute_dist_default - 2 * pi * r), involute_dist_default)
    return difference(involute_dist - s, disk(r))

intersection(
    planar_involute(1, 0.5),
    cube_centered((200, 200, 5))
)
```
I didn't bother to extend this beyond the plane, but it should be doable. I think this is novel, which is fun. I also think I may have over-complicated it a little and had to compensate with that absolute value at the end. The surface looks like this:
![Involute_sprial_surface](https://github.com/libfive/libfive/assets/13378502/2043c94c-d102-46b8-a156-a927c9b6646b)
